### ISSUE TITLE
Clasificacion de Numeros

### ISSUE THREAD

**Issue by HeladeroTV** [HeladeroTV/Tareas-IA#5]
``import numpy as np
import matplotlib.pyplot as plt
from tensorflow import keras
from tensorflow.keras import layers
from tensorflow.keras.datasets import mnist
from tensorflow.keras.utils import to_categorical

# 1. Cargar el dataset MNIST
print("Cargando dataset MNIST...")
(x_train, y_train), (x_test, y_test) = mnist.load_data()

# 2. Visualizar algunos ejemplos
plt.figure(figsize=(10, 5))
for i in range(10):
    plt.subplot(2, 5, i+1)
    plt.imshow(x_train[i], cmap='gray')
    plt.title(f"Etiqueta: {y_train[i]}")
    plt.axis('off')
plt.tight_layout()
plt.savefig('ejemplos_mnist.png')
plt.close()

# 3. Preprocesar los datos
# Normalizar los valores de píxeles a [0, 1]
x_train = x_train.astype("float32") / 255
x_test = x_test.astype("float32") / 255

# Añadir una dimensión para el canal (requerido por la red neuronal)
x_train = np.expand_dims(x_train, -1)
x_test = np.expand_dims(x_test, -1)

# Convertir etiquetas a one-hot encoding
y_train = to_categorical(y_train, 10)
y_test = to_categorical(y_test, 10)

print(f"Forma de datos de entrenamiento: {x_train.shape}")
print(f"Forma de etiquetas de entrenamiento: {y_train.shape}")
print(f"Forma de datos de prueba: {x_test.shape}")
print(f"Forma de etiquetas de prueba: {y_test.shape}")

# 4. Construir el modelo de red neuronal
modelo = keras.Sequential([
    # Aplanar la imagen 28x28 a un vector de 784 elementos
    layers.Flatten(input_shape=(28, 28, 1)),
    
    # Capa oculta con 128 neuronas y activación ReLU
    layers.Dense(128, activation="relu"),
    
    # Capa de salida con 10 neuronas (una por dígito) y activación softmax
    layers.Dense(10, activation="softmax")
])

# 5. Compilar el modelo
modelo.compile(
    optimizer="adam",
    loss="categorical_crossentropy",
    metrics=["accuracy"]
)

# Resumen del modelo
modelo.summary()

# 6. Entrenar el modelo
print("\nEntrenando el modelo...")
historial = modelo.fit(
    x_train, y_train,
    batch_size=128,
    epochs=5,
    validation_split=0.1,
    verbose=1
)

# 7. Evaluar el modelo
print("\nEvaluando el modelo...")
resultados = modelo.evaluate(x_test, y_test, verbose=2)
print(f"Precisión en datos de prueba: {resultados[1]:.4f}")

# 8. Visualizar el historial de entrenamiento
plt.figure(figsize=(12, 4))

# Gráfico de pérdida
plt.subplot(1, 2, 1)
plt.plot(historial.history['loss'], label='Entrenamiento')
plt.plot(historial.history['val_loss'], label='Validación')
plt.title('Pérdida durante el entrenamiento')
plt.xlabel('Época')
plt.ylabel('Pérdida')
plt.legend()

# Gráfico de precisión
plt.subplot(1, 2, 2)
plt.plot(historial.history['accuracy'], label='Entrenamiento')
plt.plot(historial.history['val_accuracy'], label='Validación')
plt.title('Precisión durante el entrenamiento')
plt.xlabel('Época')
plt.ylabel('Precisión')
plt.legend()

plt.tight_layout()
plt.savefig('rendimiento_modelo.png')
plt.close()

# 9. Hacer algunas predicciones y mostrarlas
num_ejemplos = 10
indices_aleatorios = np.random.randint(0, x_test.shape[0], num_ejemplos)
imagenes_test = x_test[indices_aleatorios]
predicciones = modelo.predict(imagenes_test)
etiquetas_reales = np.argmax(y_test[indices_aleatorios], axis=1)

plt.figure(figsize=(15, 5))
for i in range(num_ejemplos):
    plt.subplot(2, 5, i+1)
    plt.imshow(np.squeeze(imagenes_test[i]), cmap='gray')
    etiqueta_predicha = np.argmax(predicciones[i])
    color = 'green' if etiqueta_predicha == etiquetas_reales[i] else 'red'
    plt.title(f"Pred: {etiqueta_predicha}, Real: {etiquetas_reales[i]}", color=color)
    plt.axis('off')
plt.tight_layout()
plt.savefig('ejemplos_predicciones.png')
plt.close()

# 10. Guardar el modelo entrenado
modelo.save("modelo_mnist.h5")
print("\nModelo guardado como 'modelo_mnist.h5'")

# 11. Matriz de confusión para visualizar errores comunes
from sklearn.metrics import confusion_matrix
import seaborn as sns

# Obtener predicciones para todo el conjunto de prueba
y_pred = modelo.predict(x_test)
y_pred_clases = np.argmax(y_pred, axis=1)
y_true_clases = np.argmax(y_test, axis=1)

# Calcular la matriz de confusión
conf_mat = confusion_matrix(y_true_clases, y_pred_clases)

# Visualizar la matriz de confusión
plt.figure(figsize=(10, 8))
sns.heatmap(conf_mat, annot=True, fmt='d', cmap='Blues')
plt.title('Matriz de Confusión')
plt.xlabel('Predicción')
plt.ylabel('Real')
plt.savefig('matriz_confusion.png')
plt.close()

print("\n¡Clasificación de dígitos MNIST completada con éxito!")
print(f"Precisión final del modelo: {resultados[1]*100:.2f}%")``

Este codigo utiliza imagenes y numeros a la vez de entrenamiento normalizando los valores de pixeles y poniendo las etiquetas
Creamos una red neuronal con una capa de aplanamiento ademas de una capa oculta con 128 neuronas y una capa de salida con 10 neuronas una para cada digito donde despues evaluamos el rendimiento al final guardamos el modelo entrenado para un uso futuro
